feat: build and release docker images using github actions and container registry

maybeanerd opened the pull request:
## Related Ticket(s)
- Fixes #4375 

## Short roundup of the initial problem
There are no officially published images for the server. This makes setting up a server using docker more painful than it needs to be

## What will change with this Pull Request?
a new pipeline using github actions was added. this pipeline does two things:
1. on pull requests, builds the image without publishing. This makes sure the dockerfile still can be built. Example run: https://github.com/maybeanerd/cockatrice/actions/runs/14296665254/job/40064716966
2. on tags that contain "Release", builds the image and releases it as a package of this repository on the github container registry. this allows others to easily pull and use the images. New images will be tagged with the git tag (so e.g. Cockatrice/Cockatrice:2025-04-03-Release-2.10.2 and Cockatrice/Cockatrice:latest ) . Example run: https://github.com/maybeanerd/cockatrice/actions/runs/14296614709/job/40064609872

Builds are made for both amd64 and arm64 architecture, tagging the images respectively.

### Notes and considerations

_Naming after the repository makes GitHub know its related to this repo. This does mean however, that the image is not called "servatrice", and instead cockatrice. If this is an issue, I think we could hardcode the name, but then GitHub would not list it in the "packages" section of the repository automatically, AFAIK. not entirely sure what would be necessary for that to happen, as that's something I didn't need to do in past implementations._
**See next comment in this PR**

The pipeline seems to take quite some time. I hope caching of the containers layers will speed this up in the future, but it might be worth considering to not run them in PRs. I personally would like to ensure the image always builds successfully, but that's a tradeoff one could make.

it could probably also be enriched to release beta images, but I wanted to first get something working out without doing too much at once.


## Screenshots
Image how it will be listed in the registry, linked to the repo and with version tags
![Screenshot 2025-04-07 at 00 17 17](https://github.com/user-attachments/assets/d9eb4603-0b8a-4d36-a8a9-dc4a227b4556)

Packages section in the repos main page
![Screenshot 2025-04-07 at 00 18 40](https://github.com/user-attachments/assets/9e8c5c1e-d113-4887-a548-546ae3582d90)

maybeanerd committed:
feat: build and release docker images using github cicd

maybeanerd committed:
fix: attempt to publish to specific image name

maybeanerd commented:
I just tested the image, and the name needs to be lowercase. Since that would not be the case with this repos name, I just force named it to be cockatrice/servatrice, but this is something I could not test in my fork. we'd need to merge and wait for the next release to fully confirm that name works.

maybeanerd commented:
Looks like the pipeline in this PR already benefitted from caching, it took < 1min when the first one took 1h+

tooomm commented:
I'm not using Docker or our Dockerfile, so I can not judge if it works or how recent the notes in the [wiki](https://github.com/Cockatrice/Cockatrice/wiki/Setting-up-Servatrice#using-docker) are.

> Looks like the pipeline in this PR already benefitted from caching, it took < 1min when the first one took 1h+

Yes, I did rerun the job after it finished successfully to test the caching:
- First run did take a really long `1h 12m` for the `build and push` step.
- Second run hit the cache and it was only `2s` for `build and push`.

I'm just not sure how useful the cache will actually be or if any small code change will fully invalidate the cache and we have to always build from scratch again.
Docker cache is working in layers depending on how the Dockerfile is structured, but I would not know how that exactly works or if we can improve that for us to keep more layers reusable.

tooomm commented on .github/workflows/docker-release.yml:
```diff
@@ -0,0 +1,44 @@
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Docker meta
+        id: metal
```
`meta`?

maybeanerd committed:
fix: typo in pipeline step

maybeanerd commented:
layer 20 is the only layer that actually builds the code, so all previous layers should correctly be cached. still that layer is prob what takes the most time, so I think caching will still not help _that_ much. If you want to, we can try including it in PRs, and if it annoys you, you can simply remove the PR trigger of the pipeline

maybeanerd commented:
Concerning running the container: I was able to run it, although I did not connect it to a DB and actually used it. In any case, we didn't change anything about the image itself - just building it in the pipeline instead of manually on each machine. There shouldTM not be a difference in the resulting images.

maybeanerd commented:
If you know any users who host/run their own servers, we could ask them to try using the image I built in my fork to confirm if it works?

tooomm commented on .github/workflows/docker-release.yml:
```diff
@@ -0,0 +1,44 @@
+          platforms: linux/amd64,linux/arm64
+          push: ${{ github.event_name != 'pull_request' }}
+          tags: ${{ steps.metal.outputs.tags }}
+          labels: ${{ steps.metal.outputs.labels }}
```
```suggestion
          labels: ${{ steps.meta.outputs.labels }}
```

tooomm committed:
typo

maybeanerd committed:
typo

tooomm commented:
>If you want to, we can try including it in PRs, and if it annoys you, you can simply remove the PR trigger of the pipeline

You have it enabled for PR's to master already?
What about limiting it to ones that change files related to Servatrice (and this workflow) only?

>Concerning running the container: I was able to run it, although I did not connect it to a DB and actually use it with clients. In any case, we didn't change anything about the image itself - just building it in the pipeline instead of manually on each machine. There shouldTM not be a difference in the resulting images.

Did you try the docker-compose file for the database or the instructions in the [wiki](https://github.com/Cockatrice/Cockatrice/wiki/Setting-up-Servatrice#using-docker)?
I would not know if the image is actively used and actually works.

>If you know any users who host/run their own servers, we could ask them to try using the image I built in my fork to confirm if it works?

Maybe worth asking on the Discord?

<br><br>

I realised that the image title is listed as `Cockatrice` in the build inputs, shouldn't that be `Servatrice` or `Cockatrice Server (Servatrice)`? Description should also mention the server part.
```
labels:
  - org.opencontainers.image.created=2025-04-12T08:59:10.456Z
  - org.opencontainers.image.description=A cross-platform virtual tabletop for multiplayer card games
  - org.opencontainers.image.licenses=GPL-2.0
  - org.opencontainers.image.revision=819bf3c96b3692b64948900f68af757dffae2fad
  - org.opencontainers.image.source=https://github.com/Cockatrice/Cockatrice
  - org.opencontainers.image.title=Cockatrice
  - org.opencontainers.image.url=https://github.com/Cockatrice/Cockatrice
  - org.opencontainers.image.version=pr-5807
```
https://github.com/Cockatrice/Cockatrice/actions/runs/14388833047#summary-40437276940

<br>

Can you adjust the naming to match the style/capitalization of the other workflows better?
e.g. `Build Docker Container / AMD64 + ARM64`
![image](https://github.com/user-attachments/assets/abaa183f-0f51-4de2-b115-91c71cdc28e0)

maybeanerd commented:
Will do! I'm currently traveling though, so I'll probably not get to it the next few days

tooomm committed:
limit to certain paths for PRs & naming

tooomm commented:
Sure, no worries!
I implemented the path limitations and structured/styled the file similar to our other workflows already. It now also builds on master so we have a check+fresh cache before releases are tagged. Please have a look.

Regarding the image name we would need to see once merged how it shows and if it's linked to the repo.

Looking at the build inputs again, we probably also want to put a link to the website instead the GitHub repo for `url`: https://cockatrice.github.io/
But not sure how this information is used or where they are displayed.

Another idea:
We could upload the resulting image as artifact to the workflow run (for PR's). It can then be downloaded and tested easily.

maybeanerd commented:
> We could upload the resulting image as artifact to the workflow run (for PR's). It can then be downloaded and tested easily.

Not sure how useful this will be. unless we release them on the repository, I don't think there's a nice way to pull and use them. That might be something though, releasing experimental versions from a manual trigger in the PR pipeline? TBH, I think that's a bit overengineered though (at least, for now)

maybeanerd committed:
ci: configure image title and url

maybeanerd committed:
docker: include only necessary files and directories

this should make caching more powerful

maybeanerd commented:
> The docker caching is a bit mysterious to me still.
The new run was taking very long again, despite there was only a single small changes to master (only concerning Oracle) in the meantime.
Can we limit the data somehow to folders that are actually relevant for building Servatrice?

Definitely! what exaclty are the necessary files, only the /servatrice folder, or outside ones as well? I tried something with a best guess, but I think having your conclusive input here to limit what the image includes will be helpful

maybeanerd committed:
docker: reorder COPY with best guess of what changes least

maybeanerd committed:
build(docker): remove seemingly unnecessary files

maybeanerd commented on .github/workflows/docker-release.yml:
```diff
@@ -0,0 +1,66 @@
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to GitHub Container Registry
+        if: github.event_name != 'pull_request'
```
We should be able to use github.ref (https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#github-context) for this. If it starts with refs/tags/ it was triggered on a tag

tooomm replied:
`github.ref_type` = tag should also do

maybeanerd committed:
fix: clean up docker metadata

remove annotations, it seems they're applied from the labels already, add description

maybeanerd commented:
> It now also builds on master so we have a check+fresh cache before releases are tagged.

This will make the pipeline a bit more messy (as noted in my comments above, where we need to detect if the trigger was a tag)

Is it common for things to be merged without a PR? If not, then we should already always have a valid cache for everything that gets merged from the PR pipeline, right?

But, if you insist on also running it on the main branch, i'll try adjusting the trigger conditions for publish - just want to make it's actually useful to add that extra complexity

maybeanerd committed:
fix(ci): add back docker image annotations

maybeanerd commented:
looks like annotations were added seperately from the labels, so I readded them (and included description there as well)

maybeanerd committed:
Merge branch 'master' into feat/release-docker-images

tooomm committed:
Update desktop-build.yml

tooomm committed:
Update desktop-lint.yml

tooomm committed:
Update desktop-build.yml

tooomm committed:
Update docker-release.yml

maybeanerd committed:
fix: remove run on master and add affected files to PR trigger

maybeanerd commented:
As suggested in my longer comment, I roved the trigger from main, as PRs should already create the cache we need for release

I also added the files the dockerfile relies on to the PR trigger paths

tooomm commented on .github/workflows/docker-release.yml:
```diff
@@ -0,0 +1,71 @@
+      - '.github/workflows/docker-release.yml'
+      - 'CMakeLists.txt'
+      - 'LICENSE'
+      - 'README.md'
```
I do not think we should run the workflow for changes to license and readme file.

They barely ever change and only having them here for the cache is maybe more confusing than helpful. It should trigger on relevant changes so we test if the image still builds, but not only to keep the cache relevant.
I would much rather run generally on pushes to master (each merge) then.

maybeanerd replied:
removed them!

tooomm reviewed (approved):
Thanks for testing and confirming that the compose file with db works for you!

I do like your additions to the Dockerfile and limiting the files that are considered. 👍 
Good comments and thoughts you noted in general - makes it easy to follow along your changes when having a look some days later.

I saw you removed your note in the README on how to use the deployed Docker image.
It should be in the wiki on how to use Servatrice at least?

Other than that, I'm fine with your recent changes. See my comments.
Guess we would need to test and see the result live. :)

tooomm committed:
metadata

maybeanerd committed:
ci: run pipeline on main

this will ensure the container can always build and keep caches ready for release. push should only happen on tag triggers

It also removes some files from the PR trigger that should never break the build, and would just invalidate cache.

maybeanerd commented:
I agree with your comments, and adjusted the pipeline to run on master again, and instead push based on the ref type. Once you run the actions on this PR again, we should see if that worked as expected, or accidentally released an image 💯

tooomm committed:
Update docker-release.yml

tooomm commented:
> I agree with your comments, and adjusted the pipeline to run on master again, and instead push based on the ref type. Once you run the actions on this PR again, we should see if that worked as expected, or accidentally released an image 💯

Looking good, I think we can try it.

@ZeldaZach @ebbit1q Any objections?

ZeldaZach commented:
Thanks for taking the time to add this! Lets give it a shot :)

tooomm commented:
I opened a new PR to add a CI badge for easy monitoring and displaying the build status (#5907).
For the outstanding wiki updates, we might want to wait for a full release first to ensure building works and we can link to the GHCR image correctly.

There is another related ticket (https://github.com/Cockatrice/Cockatrice/issues/4272), do you think "testing" the image to deploy a container makes sense and is feasible?
With an Servatrice Docker image, a pre-configured docker compose setup should be possible, too.

tooomm commented:
Jep, can confirm - package is listed as `private` currently:
<img width="494" height="157" alt="image" src="https://github.com/user-attachments/assets/0e21f05b-3bc7-460b-82ac-9fbc1dc53571" />

@ebbit1q can also help.

Docs:
https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility#configuring-visibility-of-packages-for-an-organization

ebbit1q commented:
I can't, only zach has full admin rights over the org, I just have admin to this repo

@ZeldaZach please go here, scroll down and hit change visibility https://github.com/orgs/Cockatrice/packages/container/servatrice/settings

maybeanerd commented:
using the following compose (taking the one from the repo, adjusting the image source) I got the server started up no problem

<img width="3439" height="1326" alt="Screenshot 2026-03-18 at 17 43 56" src="https://github.com/user-attachments/assets/99704d67-b6c7-4c2a-8453-a86a0cce282c" />


```yaml
version: "3"

services:
  mysql:
    image: mysql
    command: --sql_mode=
    environment:
      - MYSQL_ROOT_PASSWORD=root-password
      - MYSQL_DATABASE=servatrice
      - MYSQL_USER=servatrice
      - MYSQL_PASSWORD=password
    volumes:
      - $PWD/servatrice/servatrice.sql:/docker-entrypoint-initdb.d/servatrice.sql

  servatrice:
    image: ghcr.io/cockatrice/servatrice:latest
    depends_on:
      - mysql
    ports:
      - "4748:4748"
    entrypoint: "/bin/bash -c 'sleep 10; servatrice --config /tmp/servatrice.ini --log-to-console'"
    restart: always
    volumes:
      - $PWD/servatrice/docker/servatrice-docker.ini:/tmp/servatrice.ini

```

tooomm commented:
Ok, great. Thanks for confirming.
Connecting and playing worked also fine?

It seems you exchanged
```
  servatrice:
    build:
      context: .
      dockerfile: Dockerfile
    image: servatrice
    depends_on:
      - mysql
```
with this?
```
  servatrice:
    image: ghcr.io/cockatrice/servatrice:latest
    depends_on:
      - mysql
```

tooomm commented:
Hi @maybeanerd, we just released a new stable release of Cockatrice ([v3](https://github.com/Cockatrice/Cockatrice/releases/tag/2026-05-08-Release-3.0.0)) some days back.
The automated building and publishing of the [v3 Docker image for Servatrice](https://github.com/Cockatrice/Cockatrice/pkgs/container/servatrice/850637186?tag=2026-05-08-Release-3.0.0) is working nicely, too.

Having updated docs for users would be great now! I like your suggestion from above. 👍
